# Examining Racial Discrimination in the US Job Market

In [1]:
# Import required packages
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set()

In [4]:
data = pd.read_stata('data/us_job_market_discrimination.dta')

In [5]:
#Inspect the state of the dataframe
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4870 entries, 0 to 4869
Data columns (total 65 columns):
id                    4870 non-null object
ad                    4870 non-null object
education             4870 non-null int8
ofjobs                4870 non-null int8
yearsexp              4870 non-null int8
honors                4870 non-null int8
volunteer             4870 non-null int8
military              4870 non-null int8
empholes              4870 non-null int8
occupspecific         4870 non-null int16
occupbroad            4870 non-null int8
workinschool          4870 non-null int8
email                 4870 non-null int8
computerskills        4870 non-null int8
specialskills         4870 non-null int8
firstname             4870 non-null object
sex                   4870 non-null object
race                  4870 non-null object
h                     4870 non-null float32
l                     4870 non-null float32
call                  4870 non-null float32
city        

In [9]:
data.describe()

education       ofjobs     yearsexp       honors    volunteer  \
count  4870.000000  4870.000000  4870.000000  4870.000000  4870.000000   
mean      3.618480     3.661396     7.842916     0.052772     0.411499   
std       0.714997     1.219126     5.044612     0.223601     0.492156   
min       0.000000     1.000000     1.000000     0.000000     0.000000   
25%       3.000000     3.000000     5.000000     0.000000     0.000000   
50%       4.000000     4.000000     6.000000     0.000000     0.000000   
75%       4.000000     4.000000     9.000000     0.000000     1.000000   
max       4.000000     7.000000    44.000000     1.000000     1.000000   

          military     empholes  occupspecific   occupbroad  workinschool  \
count  4870.000000  4870.000000    4870.000000  4870.000000   4870.000000   
mean      0.097125     0.448049     215.637782     3.481520      0.559548   
std       0.296159     0.497345     148.127551     2.038036      0.496492   
min       0.000000     0.000000       7.000000     1.000000      0.000000   
25%       0.000000     0.000000      27.000000     1.000000      0.000000   
50%       0.000000     0.000000     267.000000     4.000000      1.000000   
75%       0.000000     1.000000     313.000000     6.000000      1.000000   
max       1.000000     1.000000     903.000000     6.000000      1.000000   

             email  computerskills  specialskills            h            l  \
count  4870.000000     4870.000000    4870.000000  4870.000000  4870.000000   
mean      0.479261        0.820534       0.328747     0.502259     0.497741   
std       0.499621        0.383782       0.469806     0.500051     0.500051   
min       0.000000        0.000000       0.000000     0.000000     0.000000   
25%       0.000000        1.000000       0.000000     0.000000     0.000000   
50%       0.000000        1.000000       0.000000     1.000000     0.000000   
75%       1.000000        1.000000       1.000000     1.000000     1.000000   
max       1.000000        1.000000       1.000000     1.000000     1.000000   

              call         adid    fracblack    fracwhite    lmedhhinc  \
count  4870.000000  4870.000000  4784.000000  4784.000000  4784.000000   
mean      0.080493   651.777832     0.310831     0.542772    10.147275   
std       0.272079   388.690582     0.332473     0.329467     0.345780   
min       0.000000     1.000000     0.000000     0.004814     8.841738   
25%       0.000000   306.250000     0.045275     0.252164     9.965053   
50%       0.000000   647.000000     0.159950     0.571833    10.144078   
75%       0.000000   979.750000     0.516854     0.873805    10.342871   
max       1.000000  1344.000000     0.992043     0.981653    11.119290   

       fracdropout     fraccolp         linc          col          eoe  \
count  4784.000000  4784.000000  4784.000000  4870.000000  4870.000000   
mean      0.185674     0.213816     9.550801     0.719507     0.291170   
std       0.081747     0.169305     0.557097     0.449287     0.454347   
min       0.000000     0.030847     8.507345     0.000000     0.000000   
25%       0.139711     0.092559     9.220489     0.000000     0.000000   
50%       0.190751     0.145053     9.438432     1.000000     0.000000   
75%       0.238196     0.284315     9.668208     1.000000     1.000000   
max       0.356164     0.780124    11.078660     1.000000     1.000000   

       parent_sales     parent_emp  branch_sales    branch_emp          fed  \
count   1672.000000    1722.000000    608.000000    658.000000  3102.000000   
mean     587.686035    2287.051025    196.050522    755.416992     0.114765   
std     2907.629395    8902.843750    896.510864   1665.165039     0.318791   
min        0.000000       0.000000      0.000000      0.000000     0.000000   
25%       12.975000      98.000000     13.000000     97.000000     0.000000   
50%       33.350000     220.000000     34.900000    200.000000     0.000000   
75%      133.099998     700.000000     86.699997 

In [6]:
# Set options to inspect all of the columns
pd.options.display.max_columns = 65 

# Inspect the structure of the dataframe
data.head()

id ad  education  ofjobs  yearsexp  honors  volunteer  military  empholes  \
0  b  1          4       2         6       0          0         0         1   
1  b  1          3       3         6       0          1         1         0   
2  b  1          4       1         6       0          0         0         0   
3  b  1          3       4         6       0          1         0         1   
4  b  1          3       3        22       0          0         0         0   

   occupspecific  occupbroad  workinschool  email  computerskills  \
0             17           1             0      0               1   
1            316           6             1      1               1   
2             19           1             1      0               1   
3            313           5             0      1               1   
4            313           5             1      1               1   

   specialskills firstname sex race    h    l  call city kind   adid  \
0              0   Allison   f    w  0.0  1.0   0.0    c    a  384.0   
1              0   Kristen   f    w  1.0  0.0   0.0    c    a  384.0   
2              0   Lakisha   f    b  0.0  1.0   0.0    c    a  384.0   
3              1   Latonya   f    b  1.0  0.0   0.0    c    a  384.0   
4              0    Carrie   f    w  1.0  0.0   0.0    c    a  385.0   

   fracblack  fracwhite  lmedhhinc  fracdropout  fraccolp       linc  col  \
0   0.989360   0.005500   9.527484     0.274151  0.037662   8.706325  1.0   
1   0.080736   0.888374  10.408828     0.233687  0.087285   9.532859  0.0   
2   0.104301   0.837370  10.466754     0.101335  0.591695  10.540329  1.0   
3   0.336165   0.637370  10.431908     0.108848  0.406576  10.412141  0.0   
4   0.397595   0.180196   9.876219     0.312873  0.030847   8.728264  0.0   

  expminreq schoolreq  eoe  parent_sales  parent_emp  branch_sales  \
0         5            1.0           NaN         NaN           NaN   
1         5            1.0           NaN         NaN           NaN   
2         5            1.0           NaN         NaN           NaN   
3         5            1.0           NaN         NaN           NaN   
4      some            1.0           9.4       143.0           9.4   

   branch_emp  fed  fracblack_empzip  fracwhite_empzip  lmedhhinc_empzip  \
0         NaN  NaN               NaN               NaN               NaN   
1         NaN  NaN               NaN               NaN               NaN   
2         NaN  NaN               NaN               NaN               NaN   
3         NaN  NaN               NaN               NaN               NaN   
4       143.0  0.0          0.204764          0.727046         10.619399   

   fracdropout_empzip  fraccolp_empzip  linc_empzip  manager  supervisor  \
0                 NaN              NaN          NaN      0.0         1.0   
1                 NaN              NaN          NaN      0.0         1.0   
2                 NaN              NaN          NaN      0.0         1.0   
3                 NaN              NaN          NaN      0.0         1.0   
4            0.070493         0.369903    10.007352      0.0         0.0   

   secretary  offsupport  salesrep  retailsales  req  expreq  comreq  educreq  \
0        0.0         0.0       0.0          0.0  1.0     1.0     0.0      0.0   
1        0.0         0.0       0.0          0.0  1.0     1.0     0.0      0.0   
2        0.0         0.0       0.0          0.0  1.0     1.0     0.0      0.0   
3        0.0         0.0       0.0          0.0  1.0     1.0     0.0      0.0   
4        1.0         0.0       0.0          0.0  1.0     1.0     0.0      0.0   

   compreq  orgreq  manuf  transcom  bankreal  trade  busservice  othservice  \
0      1.0     0.0    1.0       0.0       0.0    0.0         0.0         0.0   
1      1.0     0.0    1.0       0.0       0.0    0.0         0.0         0.0   
2      1.0     0.0    1.0       0.0       0.0    0.0         0.0         0.0   
3      1.0     0.0    1.0       0.0       0.0    0.0         0.0         0

In [8]:
# Check for duplicate observations:
data.duplicated().sum()

0

In [7]:
# Check for missing values in race and call columns.
data[['race','call']].isnull().sum()

race    0
call    0
dtype: int64

#### There is no null value and duplicate records on call and race columns. The dataframe looks to be in good shape for the analysis.

## 1. What test is appropriate for this problem? Does CLT apply?

In [13]:
# Create a frequency table for callback status by race:
callback_table = pd.crosstab(index=data.call, columns=data.race, margins=True)

# Label the index column:
callback_table.index=['No','Yes', 'Total']

# Label the header row:
callback_table.columns = ['Black', 'White', 'Total']

**Callbacks for Black and White Sounding Name**

In [14]:
callback_table

Black  White  Total
No      2278   2200   4478
Yes      157    235    392
Total   2435   2435   4870

Another approach to reach the same conclusion is shown below.

In [15]:
# Sample size for black-sounding names
n_black = len(data[data.race == 'b'])
print("The number of black-sounding resumes is", str(len(data[data.race == 'b'])))

# number of callbacks for black-sounding names
print("The number of callbacks for black-sounding names is", str(sum(data[data.race=='b'].call)))

print('\n') 

# Sample size for white-sounding names
n_white = len(data[data.race == 'w'])
print("The number of white-sounding resumes is", str(len(data[data.race == 'b'])))

# number of callbacks for white-sounding names
print("The number of callbacks for white-sounding names is", str(sum(data[data.race=='w'].call)))

The number of black-sounding resumes is 2435
The number of callbacks for black-sounding names is 157.0


The number of white-sounding resumes is 2435
The number of callbacks for white-sounding names is 235.0


#### Conclusion:

* To figure out the 'black' and 'white' resume amounts and call backs we grouped by race and call. Both has 2435 resumes, but black got 157 calls meanwhile white got 235 calls back. 


* Since the rate of callbacks for the two groups is compared, a two-proportion z-test is appropriate.


* Central Limit Theorem (CLT) states that regardless of the population, the distribution of sample averages tends to be normal. This holds for sample sizes greater than or equal to 30. Since the size of the data is big enough and 'black' and 'white' values in race are assigned randomly to the resumes when presented to the employer, CLT could be applied here assuming that samples are respresenting the population.

## 2. What are the null and alternate hypotheses?

As observed above, it is obvious that white-sounding names received more callbacks than black-sounding names, but is the difference significant? Here is the important point to define the null and alternative hypothesis: 

- If testing something assumed to be true, the null hypothesis can reflect the assumption. 

- if testing a claim we want to be true, but can't assume, we test its opposite.

In the context of provided background information, the alternative hypothesis assumes that racial discrimination is pervasive enough to be a factor in the job market.

Null Hypothesis: There is no difference in the proportion of callbacks for white-sounding names and the proportion of callbacks for black-sounding names:

$$H_0: \hat p_{white\_callbacks} - \hat p_{black\_callbacks} = 0$$

Alternative Hypothesis: There is a significant difference in the proportion of callbacks for white-sounding names and the proportion of callbacks for black-sounding names:

$$H_a: {\hat p_{white\_callbacks}} - {\hat p_{black\_callbacks}} \neq 0$$
Significance Level: 95% Confidence:

## 3. Compute margin of error, confidence interval, and p-value. Try using both the bootstrapping and the frequentist statistical approaches.

###  a.  Bootstrapping Statistical Approach

- Out of curiosity, I used two different approach for bootstrapping, which gave me the same result. I used the function in the first one as you may see below. 

In [16]:
# Define difference function to compute difference in proportions
def diff_of_p(x,y):
    return abs(np.sum(x) - np.sum(y))

In [ ]:
# Define bootstrap function that randomly split data
def bs_split_data(data, funct, size=1):
    
    # Initialize replicates
    bs_replicates = np.empty(size)
    
    for i in range(size):
        # Permutate the concatenated array
        permutated_data= np.random.permutation(data)
        
        # Split the permutated array into two
        half_count = int(len(data)/2)
        perm_sample_1 = permutated_data[:half_count]
        perm_sample_2 = permutated_data[half_count:]
        
        # Call function
        bs_replicates[i] = func(perm_sample_1,)

In [17]:
# Define bootstrap function that randomly split data
def bs_split_data(data, func, size=1):
    
    # Initialize replicates
    bs_replicates = np.empty(size)
    
    for i in range(size):
        # Permutate the concatenated array: permutated_data
        permuted_data = np.random.permutation(data)
        
        # Split the permuted array into two: perm_sample_1, perm_sample_2
        # Assume data can be split into even numbers
        half_count = int(len(data)/2)
        perm_sample_1 = permuted_data[:half_count]
        perm_sample_2 = permuted_data[half_count:]
        
        # Call function
        bs_replicates[i] = func(perm_sample_1/half_count, perm_sample_2/half_count)
        
    return bs_replicates

In [20]:
# Defining variables
b_call_mean = b.call.mean()
w_call_mean = w.call.mean()

# Acquire 10000 bootstrap replicates using difference of proprotion
bs_replicates = bs_split_data(data.call, diff_of_p, size=10000)

bs_p_value = np.sum(bs_replicates>=diff_of_p(w_call_mean, b_call_mean))/len(bs_replicates)

print("p-value =", bs_p_value)

p-value = 0.0001


In the second bootstrapping approach, I just used for loop. 

In [23]:
# Defining variables
w = data[data.race=='w']
b = data[data.race=='b']
calls=np.sum(w.call[w.call==1])

diff_mean = w_call_mean - b_call_mean

# Construct arrays of data: white-sounding names, black-sounding names
all_callbacks = np.array([True] * int(calls) + [False] * int(len(data)-calls))

size = 10000

bs_reps_diff = np.empty(size)

for i in range(size):
    w_bs_replicates = np.sum(np.random.choice(all_callbacks, size=len(w)))
    b_bs_replicates = np.sum(np.random.choice(all_callbacks, size=len(b)))
    
    bs_reps_diff[i] = (w_bs_replicates - b_bs_replicates)/len(b)
    
bs_p_value = np.sum(bs_reps_diff >= diff_mean) / len(bs_reps_diff)

bs_ci = np.percentile(bs_reps_diff, [2.5, 97.5])
bs_mean_diff = np.mean(bs_reps_diff)

print('BOOTSTRAP RESULTS\np-value: {}\n95% conf. int.: {}'.format(bs_p_value, bs_ci))

BOOTSTRAP RESULTS
p-value: 0.0
95% conf. int.: [-0.01190965  0.01190965]


#### Conclusion:

After generating 10000 means for each group, p_value is less than 0.05. So based on this conclusion, Null Hypothesis will be rejected. The data shows that there is a racial discrimination in the US job market.

### b. Frequentist Statical Approach

In [24]:
# Assign black and white race into variables
b = data[data.race == 'b']
w = data[data.race == 'w']

# Find the mean for black and white
b_call_mean = b.call.mean()
w_call_mean = w.call.mean()
print ('Black mean: {}'.format(b_call_mean))
print ('White mean: {}'.format(w_call_mean))

# Find the difference of mean for black and white
diff_mean = w_call_mean - b_call_mean
print ('Difference of mean between white and black: {}'.format(diff_mean))

Black mean: 0.0644763857126236
White mean: 0.09650924056768417
Difference of mean between white and black: 0.03203285485506058


### z-score

In [25]:
w_call = np.sum(data[data.race=='w'].call)
b_call = np.sum(data[data.race=='b'].call)
w_ln = len(data[data.race=='w'])
b_ln = len(data[data.race=='b'])            
phat = (w_call + b_call) / (w_ln + b_ln)
z = (diff_mean) / np.sqrt(phat * (1 - phat) * (( 1/ float(w_ln)) + (1/float(b_ln))))
print ('z score:{}'.format(z))

z score:4.108412235238472


### p-value

In [26]:
p = stats.norm.cdf(-z)*2
print ('p-value:{}'.format(p))

p-value:3.983885409504418e-05


### Margin of error

In [27]:
moe = 1.96 * np.sqrt(phat * (1 - phat) * (( 1/ float(w_ln)) + (1/float(b_ln))))
print("Margin of Error: {}".format(moe))

Margin of Error: 0.015281912310894095


### Confidence Interval

In [28]:
ci = diff_mean + np.array([-1, 1]) * moe
print("Confidence interval: {}".format(ci))

Confidence interval: [0.01675094 0.04731477]


#### Conclusion:

The p-value is very low than 5% we can conclude that there is a significant difference between two groups, so the null hypothesis is rejected. There is sufficient evidence to warrant rejection of the claim that there is no difference in the proportion of callbacks for white-sounding names and the proportion of callbacks for black-sounding names. 

## 4. Write a story describing the statistical significance in the context or the original problem.

Researchers randomly assigned identical resumes to black-sounding or white-sounding names and recorded the impact on requests for interviews from employers in order to study the effect of racial discrimination in the U.S. labor market. That analysis tested the effect of race on callback success. Assuming controls were appropriate for other factors related to callbacks, the statistics test indicates that racial discrimination in the United States labor market is significant. 

Given our sample data and our findings in this project, we can infer that a black- or white-sounding name on a resume has a significant effect on requests for interviews from employers. We also see that resumes with a white-sounding name see a 50% higher callback rate over resumes with a black-sounding name. 

Based on the result of p-value which is too close to 0, we have a strong evidence to reject the null hypothesis which we assumes at the beginning as no racial discrimination when calling back to resumes. We can say that there is a difference in the proportion of callbacks for white-sounding names and the proportion of callbacks for black-sounding names when returning to resumes. In a conclusion, our findings indicate a racial discrimination in USA job market when returning to resumes. 

## 5. Does your analysis mean that race/name is the most important factor in callback success? Why or why not? If not, how would you amend your analysis?

The analysis we made does not mean that race is the most crucial factor in callback success. Despite the test shows that callback rates differs, other variables as such year of experience, education or maybe military statistics might effect the callback rates and need to be also checked. It would be prudent to examine these other possible data points as they may be responsible for the significant difference in callback rate. We explored the data in a manner that is framed by the assumption that race plays a factor in callback rate. I think it would be logical to expand the analysis over the entire dataset. With this analysis, we can proceed confirming the effect of race on callback rates or perhaps find other factors which affect callback rates.